paddle/fluid/operators/fused_elemwise_activation_op.cc

```diff
@@ -0,0 +1,225 @@
+                  "fused_elemwise_activation_grad has two methods to get the "
+                  "dx and dy, one "
+                  "is to use the 'Out', and the other is not to use it. "
+                  "The former method will save the time of recomputing the "
+                  "'Out', but it must occupy the memory to store the 'out'. "
+                  "While, the later method can avoid occupying the memory, "
+                  "but it must recompute the 'Out'. The default value is true.")
+        .SetDefault(true);
+    AddAttr<std::string>("functor_list", "The functors that should be fused.")
+        .AddCustomChecker([&](const std::string &functor_list) {
+          PADDLE_ENFORCE(ValidCheck(functor_list));
+        });
+
+    AddComment(R"DOC(
+FusedElemwiseActivation Operator.
+
+At present, FusedElemwiseActivation only supports Two kinds of compound
+operators (elementwise_op and activation_op):
+
+    Z = Binary(X, Unary(Y))
+    Z = Unary(Binary(X, Y))
+
+The attributions of activation_op can be get from fused_elemwise_activation_op's
+attributions. functor_list records the functors to be fused, for example
+"scale,elementwise_add".
+
+)DOC");
+  }
+
+ private:
+  bool ValidCheck(const std::string &functors) {
+    std::unordered_set<std::string> unary_fun = {"scale", "relu"};
+    std::unordered_set<std::string> binary_fun = {"elementwise_add"};
+
+    size_t pos = functors.find(",");
+    PADDLE_ENFORCE_NE(pos, std::string::npos);
+    auto func_1 = functors.substr(0, pos);
+    auto func_2 = functors.substr(pos + 1, functors.size());
+    std::string unary_fun_str;
+    if (binary_fun.count(func_1)) {
+      unary_fun_str = func_2;
+    } else if (binary_fun.count(func_2)) {
+      unary_fun_str = func_1;
+    } else {
+      PADDLE_THROW("%s and %s are not included in fused_list.", func_1, func_2);
+    }
+    PADDLE_ENFORCE_EQ(unary_fun.count(unary_fun_str), 1,
+                      "%s is not included in fused_list.", unary_fun_str);
+    return true;
+  }
+};
+
+class FusedElemwiseActivationGradMaker
+    : public framework::SingleGradOpDescMaker {
+ public:
+  using framework::SingleGradOpDescMaker::SingleGradOpDescMaker;
+
+ protected:
+  std::unique_ptr<framework::OpDesc> Apply() const override {
+    auto *op_desc_ptr = new framework::OpDesc();
+    op_desc_ptr->SetType(this->ForwardOpType() + "_grad");
+
+    for (auto &input_param : this->InputNames()) {
+      op_desc_ptr->SetInput(input_param, this->Input(input_param));
+      op_desc_ptr->SetOutput(framework::GradVarName(input_param),
+                             this->InputGrad(input_param, true));
+    }
+
+    for (auto &output_param : this->OutputNames()) {
+      op_desc_ptr->SetInput(output_param, this->Output(output_param));
+      op_desc_ptr->SetInput(framework::GradVarName(output_param),
+                            this->OutputGrad(output_param));
+    }
+    op_desc_ptr->SetAttrMap(this->Attrs());
```

Comment:
For fused_elemwise_activation_op, the gradient operator's input should have forward I/O.
fused_elemwise_activation_op support fusing elementwise functor and activation, the pattern is `z=f1(f2(x,y))` and `z=f1(x,f2(y))`, because we did not save the intermediate results, for example `f2(x,y)` and `f2(y)`, the input of derivative of `f1` should be recomputed.
Of course, if the pattern is `f1(x,f2(y))` and `f1` is `add`, `x` is unnecessary for the derivative of `f1`.